src/query/storage/promremote/storage_test.go

```diff
@@ -453,3 +491,87 @@ func getWriteRequest(promServer *promremotetest.TestPromServer) *prompb.WriteReq
 	}
 	return wq
 }
+
+func LoadTestPromRemoteStorage(t *testing.T, jitter bool, numTenants, numSeries, numSamples int) {
+	fakeProm := promremotetest.NewServer(t, jitter)
+	defer fakeProm.Close()
+	scope := tally.NewTestScope("test_scope", map[string]string{})
+	defer verifyMetrics(t, scope)
+	labelName := "test_tag_name"
+	labelValues := make([][]byte, numSeries)
+	for i := 0; i < numSeries; i++ {
+		labelValues[i] = []byte(fmt.Sprintf("test_tag_value_%d", i))
+	}
+	tenantRules := make([]TenantRule, numTenants)
+	for i := 0; i < numTenants; i++ {
+		filterValues, _ := filters.ValidateTagsFilter(fmt.Sprintf("%s:%s", labelName, labelValues[i%numSeries]))
+		filter, _ := filters.NewTagsFilter(filterValues, filters.Conjunction, filters.TagsFilterOptions{})
+		tenantRules[i] = TenantRule{
+			Tenant: fmt.Sprintf("tenant_%d", i),
+			Filter: filter,
+		}
+	}
+	promStorage, err := NewStorage(Options{
+		endpoints:     []EndpointOptions{{name: "testEndpoint", address: fakeProm.WriteAddr(), tenantHeader: "TENANT"}},
+		scope:         scope,
+		logger:        logger,
+		poolSize:      10,
+		queueSize:     numTenants * numSamples,
+		tenantDefault: "unknown",
+		tickDuration:  ptrDuration(tickDuration),
+		tenantRules:   tenantRules,
+	})
+	require.NoError(t, err)
+
+	totalSamples := 0
+	for i := 0; i < numSamples; i++ {
+		datapoints := make(ts.Datapoints, 0, numSeries)
+		for j := 0; j < cap(datapoints); j++ {
+			datapoints = append(datapoints, ts.Datapoint{
+				Timestamp: xtime.Now(),
+				Value:     rand.Float64(),
+			})
+		}
+		totalSamples += len(datapoints)
+		wq, _ := storage.NewWriteQuery(storage.WriteQueryOptions{
+			Tags: models.Tags{
+				Opts: models.NewTagOptions(),
+				Tags: []models.Tag{{
+					Name:  []byte(labelName),
+					Value: labelValues[rand.Intn(numSeries)],
+				}},
+			},
+			Datapoints:   datapoints,
+			Unit:         xtime.Millisecond,
+			FromIngestor: (rand.Int() % 2) == 0,
+		})
+		err := promStorage.Write(context.TODO(), wq)
```

Comment:
If you use a goroutine to write asynchronously, you don't need to make the channel large enough to buffer all writes. Perhaps, do the following. Some writes may fail while the storage is being closed, but it's closer to the real shutdown situation.
```
var wg sync.WaitGroup
...
wg.Add(1)
go func() {
   err := promStorage.Write(context.TODO(), wq)
   wg.Done()
}()
...
closeWithCheck(t, promStorage)
wg.Wait()

```

Reply:
I think the promStorage.Close() waits for all writes to finish already.

Reply:
`promStorage.Close()` doesn't wait for pending `Write()` calls which put items to the data queue channel. It closes the data queue channel. If there are concurrent writes going to the channel, those writes from `promStorage.Write()` will fail because they are writing to a closed channel. If the call site doesn't wait for all pending `Write()` to finish before calling `Close()`, those writes will fail. But it shouldn't be a problem because those write failures should be bubbled up further.

Reply:
yep, I've confirmed the compositeStore will bubble up error and RWA/obs-agent should retry

Reply:
or we can have a flag to return error to Write() when Close() is called ?